indexing-service/src/main/java/org/apache/druid/indexing/common/task/OverlordCoordinatingSegmentAllocator.java

```diff
@@ -0,0 +1,149 @@
+import org.joda.time.Interval;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+
+/**
+ * Segment allocator which allocates new segments using the overlord per request.
+ */
+public class OverlordCoordinatingSegmentAllocator implements SegmentAllocator
+{
+  private final ActionBasedSegmentAllocator internalAllocator;
+
+  OverlordCoordinatingSegmentAllocator(
+      final TaskToolbox toolbox,
+      final @Nullable SupervisorTaskAccess supervisorTaskAccess,
+      final DataSchema dataSchema,
+      final TaskLockHelper taskLockHelper,
+      final boolean appendToExisting,
+      final PartitionsSpec partitionsSpec
+  )
+  {
+    this.internalAllocator = new ActionBasedSegmentAllocator(
+        toolbox.getTaskActionClient(),
+        dataSchema,
+        (schema, row, sequenceName, previousSegmentId, skipSegmentLineageCheck) -> {
+          final GranularitySpec granularitySpec = schema.getGranularitySpec();
+          final Interval interval = granularitySpec
+              .bucketInterval(row.getTimestamp())
+              .or(granularitySpec.getSegmentGranularity().bucket(row.getTimestamp()));
+          final PartialShardSpec partialShardSpec = createShardSpecFactory(
+              appendToExisting,
+              partitionsSpec,
+              taskLockHelper,
+              interval
+          );
+          if (supervisorTaskAccess != null) {
+            return new SurrogateAction<>(
+                supervisorTaskAccess.getSupervisorTaskId(),
+                new SegmentAllocateAction(
+                    schema.getDataSource(),
+                    row.getTimestamp(),
+                    schema.getGranularitySpec().getQueryGranularity(),
+                    schema.getGranularitySpec().getSegmentGranularity(),
+                    sequenceName,
+                    previousSegmentId,
+                    skipSegmentLineageCheck,
+                    partialShardSpec,
+                    taskLockHelper.getLockGranularityToUse()
+                )
+            );
+          } else {
+            return new SegmentAllocateAction(
+                schema.getDataSource(),
+                row.getTimestamp(),
+                schema.getGranularitySpec().getQueryGranularity(),
+                schema.getGranularitySpec().getSegmentGranularity(),
+                sequenceName,
+                previousSegmentId,
+                skipSegmentLineageCheck,
+                partialShardSpec,
+                taskLockHelper.getLockGranularityToUse()
+            );
+          }
+        }
+    );
+  }
+
+  @Nullable
+  @Override
+  public SegmentIdWithShardSpec allocate(
+      InputRow row,
+      String sequenceName,
+      String previousSegmentId,
+      boolean skipSegmentLineageCheck
+  ) throws IOException
+  {
+    return internalAllocator.allocate(row, sequenceName, previousSegmentId, skipSegmentLineageCheck);
+  }
+
+  private static PartialShardSpec createShardSpecFactory(
```

Comment:
This method name can be updated to match the rename to "PartialShardSpec"

Reply:
Renamed.